aws-janitor/resources/efs.go

```diff
@@ -0,0 +1,210 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/service/efs"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	maxRetries   = 120
+	pollInterval = time.Second
+)
+
+// ElasticFileSystems: https://docs.aws.amazon.com/sdk-for-go/api/service/efs/#EFS.DescribeFileSystems
+
+type ElasticFileSystems struct{}
+
+func (ElasticFileSystems) MarkAndSweep(opts Options, set *Set) error {
+	logger := logrus.WithField("options", opts)
+	svc := efs.New(opts.Session, aws.NewConfig().WithRegion(opts.Region))
+
+	// Paged calls, defer deletion until we have the whole list.
+	var (
+		fileSystemsToDelete  []*elasticFileSystem
+		mountTargetsToDelete []*mountTarget
+	)
+
+	// Mark and sweep file systems for deletion.
+	fsPageFunc := func(page *efs.DescribeFileSystemsOutput, _ bool) bool {
+		for _, fs := range page.FileSystems {
+			f := &elasticFileSystem{
+				ID: *fs.FileSystemId,
+			}
+			if set.Mark(f) {
+				logger.Warningf("%s: deleting %T: %s", f.ARN(), fs, *fs.Name)
+				if !opts.DryRun {
+					fileSystemsToDelete = append(fileSystemsToDelete, f)
+				}
+			}
+		}
+		return true
+	}
+	if err := svc.DescribeFileSystemsPages(&efs.DescribeFileSystemsInput{}, fsPageFunc); err != nil {
+		return err
+	}
+
+	// Mark and sweep mount targets for deletion.
```

Comment:
it looks like this comment is a little misleading - it appears that this deletes all mount targets for a filesystem, rather than marking + sweeping based on age.